crates/telio-network-monitors/src/monitor.rs

```diff
@@ -0,0 +1,127 @@
+//! Module to monitor changes in network.
+//! Get notified about that network path has changed
+//! and update local IP address cache
+use once_cell::sync::Lazy;
+use std::io;
+use std::sync::{Arc, Mutex as StdMutex};
+use telio_utils::{local_interfaces, telio_log_debug, telio_log_warn, GetIFError, GetIfAddrs};
+use tokio::{sync::broadcast::Sender, task::JoinHandle};
+/// Sender to notify if there is a change in OS interface order
+pub static PATH_CHANGE_BROADCAST: Lazy<Sender<()>> = Lazy::new(|| Sender::new(2));
+/// Vector containing all local interfaces
+pub static LOCAL_ADDRS_CACHE: Lazy<Arc<StdMutex<Vec<if_addrs::Interface>>>> =
+    Lazy::new(|| Arc::new(StdMutex::new(Vec::new())));
+#[cfg(any(target_os = "macos", target_os = "ios", target_os = "tvos"))]
+static NETWORK_PATH_MONITOR_START: std::sync::Once = std::sync::Once::new();
+
+#[derive(Debug)]
+/// Struct to monitor network
+pub struct NetworkMonitor {
+    nw_path_monitor_monitor_handle: Option<JoinHandle<io::Result<()>>>,
+    // Mac and Windows use callbacks, whereas in Linux a socket is registered
+    // with netlink multicast group messages
+    #[cfg(target_os = "linux")]
+    linux_monitor_handle: Option<JoinHandle<io::Result<()>>>,
+    #[cfg(target_os = "windows")]
+    windows_monitor_handle: crate::windows::SafeHandle,
+}
+
+impl NetworkMonitor {
+    /// Sets up and spawns network monitor
+    pub async fn new<G>(if_addr: G) -> Result<NetworkMonitor, GetIFError>
+    where
+        G: GetIfAddrs + Clone,
+    {
+        if let Ok(mut guard) = LOCAL_ADDRS_CACHE.lock() {
+            *guard = local_interfaces::gather_local_interfaces(&if_addr)?;
+            telio_log_debug!("local cache {:?}", *guard);
+        }
+
+        let nw_path_monitor_monitor_handle = Some(tokio::spawn({
```

Comment:
Looking at this again if I understand correctly this is platform independent code, and not setting up `NWPathMonitor`? If that is the case what do you think about calling it something other than `nw_path_monitor`?